Fix `--milestone` flag on `issues list` command

fsmiamoto opened the pull request:
This PR adds a fix to the `--milestone` flag on the `issues list` command as described in #1441 
```
$ cli issue list --milestone "AWS Migration"

Showing 1 of 1 issue in Safecast/safecastapi that matches your search

#89  Need How To restore from backup    about 1 year ago
```

I'm not sure if the approach I took was the best one so any feedback would be great :+1:

Edit: 
Closes #1441

fsmiamoto committed:
fix searching issues by the milestone name

This fix involves using the REST API since the GraphQL one doesn't
seem to return the necessary ID for the subsequent issues query.

fsmiamoto committed:
register mocks for the new request

fsmiamoto commented:
The use of the REST API appears to be unavoidable since the GraphQL API does not provide the necessary ID for the use in the query.

Hopefully this could change in the future to avoid this workaround.

mislav commented on api/queries_issue.go:
```diff
@@ -246,8 +247,18 @@ func IssueList(client *Client, repo ghrepo.Interface, state string, labels []str
+			return nil, err
+		}
+		for i := range milestones {
+			if milestones[i].Title == milestoneString {
```
I think a case-insensitive comparison would benefit users who type out a milestone name instead of pasting it! You can use `strings.EqualFold`

mislav commented on command/issue_test.go:
```diff
@@ -170,10 +170,13 @@ func TestIssueList_tty_withFlags(t *testing.T) {
 			assert.Equal(t, "probablyCher", params["assignee"].(string))
 			assert.Equal(t, "foo", params["author"].(string))
 			assert.Equal(t, "me", params["mention"].(string))
-			assert.Equal(t, "1.x", params["milestone"].(string))
```
Instead of removing this assertion, could it verify that the correct milestone ID was passed?

mislav commented on api/queries_issue.go:
```diff
@@ -246,8 +247,18 @@ func IssueList(client *Client, repo ghrepo.Interface, state string, labels []str
+
 	if milestoneString != "" {
-		variables["milestone"] = milestoneString
+		milestones, err := RepoMilestonesREST(client, repo)
```
You can use our existing GraphQL-based `RepoMilestones()` function and read the GraphQL ID for a milestone. Then, you can base64-decode the GraphQL ID, which will get you a string like `09:Milestone12345`. The final number in that string is the database ID of the milestone.

It's a bit nasty, but will allow you to reuse the existing milestones query instead of having to write the new REST-based one!

fsmiamoto replied:
Good point :+1: 
Should we assume the the decoded string will follow a pattern like: `^\d+:Milestone(\d+)$`?

mislav replied:
Yes, I think you can safely assume that and let it error out if it doesn't.

mislav reviewed (changes requested):
Excellent; thanks for tackling this!

fsmiamoto committed:
use graphql for repo milestones on issue list

Since we can obtain the necessary ID for the query on
node ID from the GraphQL API, it makes sense to remove
the REST version.

fsmiamoto committed:
add mock for milestones on issue list

fsmiamoto committed:
improve comments

fsmiamoto committed:
change comment to refer to pr

fsmiamoto commented:
@mislav 
Thanks for the review, I think I've managed to implement the necessary changes

tierninho commented:
Thanks for the PR! I played around with this and here are some findings:

1. There should be some type of "no match" result if the following is run `gh issue list --milestone "sdhfsdhjf"`. 

   Currently it returns this:

   ```
   Showing 5 of 5 issues in tierninho/Blah that match your search

   #12  sdfsdf                          about 1 minute ago
   #10  sdfsdfsdfssfdssdf  (duplicate)  about 1 minute ago
   #5   sdfsdf                          about 6 months ago
   #4   My issue title                  about 1 minute ago
   #3   My issue title                  about 8 months ago
   ```

   it instead should show something like: 

   `No milestones match your search in tierninho/Blah`

2. The help menu should be edited from:
   `-m, --milestone name    Filter by milestone name`
   to
   `-m, --milestone string    Filter by milestone`

mislav commented on api/queries_issue.go:
```diff
@@ -246,8 +248,23 @@ func IssueList(client *Client, repo ghrepo.Interface, state string, labels []str
+			return nil, err
+		}
+
+		for i := range milestones {
```
If we were unable to find a matching milestone, this should error out with something like: `no milestone found with title: %q`

mislav reviewed (changes requested):
Almost there! One small but important change needed before we would merge this.

fsmiamoto committed:
add check for milestone not found

fsmiamoto committed:
add test for milestone not found

fsmiamoto commented:
@tierninho @mislav 
Thanks for comments! 

 I've added the check for a not found milestone, alongside with a test, and it should solve the first issue pointed out by @tierninho.

For the second one, it looks like the pattern of using `name` is spread through out the project so I suggest we make no changes.

tierninho commented:
Thanks for adding below @fsmiamoto 

> no milestone found with title: "sdhfsdhjf"

One nitpick though: please wrap it with two empty lines like the pic, as it will match current messaging. Appreciate it!

![Screen Shot 2020-08-05 at 3 26 42 PM](https://user-images.githubusercontent.com/14828183/89480298-ada69900-d730-11ea-8fe5-09b9a1f42e61.png)

mislav reviewed (approved):
Fantastic. Thank you so much!

fsmiamoto commented:
Thanks for testing it out @tierninho  :+1: 

As for the formatting, since we raise an error for not finding the milestone, I've tried to keep to stick to the existing pattern.

I definitely agree that it looks better on the second command but since it doesn't actually raise an error, the output is handled differently.

mislav committed:
Merge remote-tracking branch 'origin' into fix-milestone

mislav committed:
Add ability to pass milestone by number

vilmibm committed:
Merge remote-tracking branch 'origin/trunk' into fix-milestone

matschaffer commented:
That looks great! Nice work @fsmiamoto - love that http://github.com/safecast got an implicit mention in all this 😆
